src/cli/cli_udp.cpp

```diff
@@ -143,7 +157,20 @@ otError UdpExample::ProcessSend(Arg aArgs[])
 
     if (!aArgs[2].IsEmpty())
     {
-        SuccessOrExit(error = aArgs[0].ParseAsIp6Address(messageInfo.mPeerAddr));
+        error = aArgs[0].ParseAsIp6Address(messageInfo.mPeerAddr);
+        if (error != OT_ERROR_NONE && !aArgs[0].IsEmpty())
+        {
+            // It might be an IPv4 address, let's have a try.
+            otIp4Address ip4Address;
+
+            SuccessOrExit(otIp4AddressFromString(aArgs[0].GetCString(), &ip4Address));
+            SuccessOrExit(
+                error = otIp6AddressSynthesizeFromIp4Address(GetInstancePtr(), &ip4Address, &messageInfo.mPeerAddr));
+
+            OutputFormat("Sending to IPv4-converted IPv6 addresses: ");
+            OutputIp6AddressLine(messageInfo.mPeerAddr);
+        }
+        SuccessOrExit(error);
```

Comment:
This code pattern (parse as IP6 and if it fails try as IPv4 and synthesize) is repeated couple of times.
How about adding a method for it?